Follow XDG Base Directory Spec for config files

jamoes opened the pull request:
Modifies the GetDefaultDataDir method to comply with the XDG Base Directory Specification (https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html).
On Unix systems only:
 - If \~/.bitcoin exists, return '\~/.bitcoin` (this ensures backward compatibility)
 - If the XDG_CONFIG_HOME environment variable is set, return '$XDG_CONFIG_HOME/bitcoin'
 - Otherwise, return '~/.config/bitcoin'.

This resolves Issue #179

jamoes committed:
Follow XDG Base Directory Spec for config files

Modifies the GetDefaultDataDir method to comply with the XDG Base Directory
Specification (https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html).
On Unix systems only:
 - If ~/.bitcoin exists, return '~/.bitcoin` (this ensures backward compatibility)
 - If the XDG_CONFIG_HOME environment variable is set, return '$XDG_CONFIG_HOME/bitcoin'
 - Otherwise, return '~/.config/bitcoin'.

zander commented:
Thanks!

In Bitcoin we have a design mistake that puts the datafiles (the blockchain) in the same dir as the config file. Likely because Satoshi was a Windows person.
The change you made here is almost good, but it also moves the blockchain into the config dir.
I expected those to move to the ~/.local dir.

Would you be able to also look at the GetConfigFile() method in the same file which uses the method you changed?

zander commented:
PR passed CI on all branches and running tests.

jamoes committed:
Split data and config directories

Modifies the GetDefaultDataDir and GetConfigFile methods to comply with the
XDG Base Directory Specification. Only applies to unix systems.

New behavior for GetDefaultDataDir:
 - If ~/.bitcoin exists, return '~/.bitcoin` (this ensures backward compatibility)
 - If the XDG_DATA_HOME environment variable is set, return '$XDG_CONFIG_HOME/Bitcoin'
 - Otherwise return `~/.local/share/Bitcoin'

New behavior for GetConfigFile:
 - If <datadir>/bitcoin.conf exists, return '<datadir>/bitcoin.conf' (this ensures backward compatibility)
 - If the XDG_CONFIG_HOME environment variable is set, return '$XDG_CONFIG_HOME/Bitcoin/bitcoin.conf'
 - Otherwise return '~/.config/Bitcoin/bitcoin.conf'

Note that the directory name is now capitalized (Bitcoin). This is for
compatibility with the existing config file (~/.config/Bitcoin/Bitcoin-Qt.conf)
that is being written by the QSettings class of the gui.

zander commented:
Works great,thanks!
